# <font color='blue'><div style="text-align: center">회귀분석(Regression)</font> 

* Francis Galton's 1875 illustration of the correlation between the heights of adults and their parents. The observation that adult children's heights tended to deviate less from the mean height than their parents suggested the concept of   "regression toward the mean", giving regression its name.

<img src="https://drive.google.com/uc?id=1hEIe9tmbVxr5UlIVddtz7wdVc6XE3e_5" width="400" height="300">

* 위의 얘기는 1875년 갤튼이 부모의 키와 자식의 키의 관계 그림을 이용해 "regression toward the mean" 이라는 주장을 했다는 것이다. 

* 여기서, "regression toward the mean" 은 무슨 의미일까? 
* 이는 키가 큰 부모의 자식은 키가 조금 작아지는 경향이 있고 키가 작은 부모의 자식은 키가 조금 커지는 경향이 있어 자식의 키가 평균으로 회귀(regress)한다는 의미이다.



In [1]:
from google.colab import drive
drive.mount('/gdrive')

Mounted at /gdrive


In [2]:
import pandas as pd
from sklearn import datasets, linear_model
data = pd.read_csv("/gdrive/My Drive/ML/data/GaltonFamilies.csv")
maleData = data[data['gender'] == "male"]
femaleData = data[data['gender'] == "female"]
data[:10]

Unnamed: 0 family  father  mother  ...  children  childNum  gender childHeight
0           1    001    78.5    67.0  ...         4         1    male        73.2
1           2    001    78.5    67.0  ...         4         2  female        69.2
2           3    001    78.5    67.0  ...         4         3  female        69.0
3           4    001    78.5    67.0  ...         4         4  female        69.0
4           5    002    75.5    66.5  ...         4         1    male        73.5
5           6    002    75.5    66.5  ...         4         2    male        72.5
6           7    002    75.5    66.5  ...         4         3  female        65.5
7           8    002    75.5    66.5  ...         4         4  female        65.5
8           9    003    75.0    64.0  ...         2         1    male        71.0
9          10    003    75.0    64.0  ...         2         2  female        68.0

[10 rows x 9 columns]

In [3]:
maleData['father'].mean()

69.13762993762994

In [4]:
maleData2 = maleData[maleData['father'] > 69.14]
maleData1 = maleData[maleData['father'] <= 69.14]

Y = maleData2['childHeight'].to_numpy()
X = maleData2['father'].to_numpy().reshape(-1,1)
linreg = linear_model.LinearRegression(fit_intercept = False)
model = linreg.fit(X, Y)
coef2 = model.coef_

Y = maleData1['childHeight'].to_numpy()
X = maleData1['father'].to_numpy().reshape(-1,1)
linreg = linear_model.LinearRegression(fit_intercept = False)
model = linreg.fit(X, Y)
coef1 = model.coef_
print(coef1, coef2)

[1.01501728] [0.98617831]


* 위 코드는 아들과 아버지의 키에서 회귀식을 구하는 코드다.
* 아버지 중에 평균 키 이상인 집단에서 구한 기울기는 0.986로 1보다 작고 평균키 이하인 집단에서 구한 기울기는 1.015로 1보다 크게 측정되었다. 
* 이는 아버지가 크면 아들의 키가 약간 작아지는 경향이 있고, 아버지가 작으면 아들의 키가 약간 커지는 경향이 있다는 갤튼의 회귀법칙을 증명할 수 있는 결과다.
* 위의 실험을 엄마의 키와 딸의 키로 했을 때도 같은 결과를 얻을 수 있었다. 